### PR TITLE
[oneDPL] [test] Fix heap use after free in uninitialized_fill_destroy.pass

### PR DESCRIPTION
**Pull request by mmichel11**
In `uninitialized_fill_destroy.pass.cpp`, we currently allocate memory into a smart pointer to manage its lifetime. However, there are two issues with this: since the memory is allocated with `new` it is already initialized when it is passed to the test function, and the test object's destructor will be called a second time after the `std::destroy` test before deallocation. This has crashes in this test on certain OSs (certain test configs with MacOS) and can be observed with address sanitizers elsewhere.

This fix manually allocates the memory with `malloc` so it is passed uninitialized to the test and manually deallocates it with `free`.

### CODE REVIEW

File: test/parallel_api/memory/specialized.algorithms/uninitialized_fill_destroy.pass.cpp [hunk 1]
```diff
@@ -198,6 +198,7 @@ test_uninitialized_fill_destroy_by_type()
         invoke_on_all_policies<>()(test_destroy_n<T>(), p_begin, p_end, T(), n,
                                    ::std::is_trivial<T>());
 #endif
+        ::std::free(p_begin);
```
Comment (akukanov) [inline test/parallel_api/memory/specialized.algorithms/uninitialized_fill_destroy.pass.cpp:201]: `free` should be guarded by the same preprocessor conditions as `malloc`.

Comment (mmichel11) [reply test/parallel_api/memory/specialized.algorithms/uninitialized_fill_destroy.pass.cpp]: I have switched to using a custom deleter with `std::unique_ptr` that frees the allocation.

**Comment by akukanov** [comment]
Perhaps it makes sense to add a comment why malloc/free is used here.

**Comment by mmichel11** [comment]
> Perhaps it makes sense to add a comment why malloc/free is used here.

I agree. I have added a comment explaining why it is used.

**Comment by MikeDvorskiy** [comment]
> > Perhaps it makes sense to add a comment why malloc/free is used here.
> 
> I agree. I have added a comment explaining why it is used.

I think the added comment doesn't  quite explain why we use "malloc / free" instead of a typical approach with T[] and smart pointer which  calls delete[] within its destructor..

I would suggest re-writting comment like this:

`We use malloc / free instead of a typical approach with operator new T[] and a smart pointer (which  calls operator delete[] within its destructor) to avoid double creation/destroy the elements of type Tin the test.`

**Review by MikeDvorskiy — APPROVED** [review]
LGTM

**Pull request merged**